### PR TITLE
Add CLI with Cliffy 

### PR DESCRIPTION
**Pull request by mllnd**

### CODE REVIEW

File: bin/withdrawer.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import { Command } from "https://deno.land/x/cliffy@v1.0.0-rc.1/command/mod.ts";
+import { Input, prompt, Select } from "https://deno.land/x/cliffy@v1.0.0-rc.1/prompt/mod.ts";
+
+let withdrawerPrompt: {
+    provider: string;
```
Comment (mikroskeem) [inline bin/withdrawer.ts:5]:
```suggestion
    provider: "vault" | "plain-text";
```

File: bin/withdrawer.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                    if (value.length < 1) {
+                        return "Path must not be empty";
+                    }
+                    return true;
+                },
+                before: async ({ provider }, next) => {
+                    if (provider === "vault") {
+                        await next();
+                    } else {
+                        await next("keypair");
```
Comment (mikroskeem) [inline bin/withdrawer.ts:40]:
```suggestion
                        await next("plain-text");
```

? Or do we have three different values?

**Review by mikroskeem — CHANGES_REQUESTED** [review]

File: bin/withdrawer.ts (hunk 2 above)
Comment (mllnd) [reply bin/withdrawer.ts:53]: N/A

File: bin/withdrawer.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+export type SendPrompt = Prompt & {
+    amount: number;
+    recipient: string;
+};
+
+let withdrawPrompt: Prompt;
+let sendPrompt: SendPrompt;
+
+const getSigner = async (prompt: Prompt | SendPrompt): Promise<RawSigner> => {
```
Comment (mikroskeem) [inline bin/withdrawer.ts:23]:
```suggestion
const getSigner = async (prompt: Prompt): Promise<RawSigner> => {
```

`SendPrompt` fields are not used here

File: bin/withdrawer.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    keypair: string | undefined;
+};
+
+export type SendPrompt = Prompt & {
+    amount: number;
+    recipient: string;
+};
+
+let withdrawPrompt: Prompt;
+let sendPrompt: SendPrompt;
```
Comment (mikroskeem) [inline bin/withdrawer.ts:21]: Why do you have those as global variables? Could pass in necessary parameters to withdraw & send functions directly. Furthermore those should be `Prompt | undefined` to be correct.

File: bin/withdrawer.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    return new RawSigner(keypair, provider);
+};
+
+const withdraw = async () => {
+    const signer = await getSigner(withdrawPrompt);
+    const tx = await withdrawStakeObjects(signer);
+    console.log(tx);
+};
+
+const send = async () => {
```
Comment (mikroskeem) [inline bin/withdrawer.ts:41]:
```suggestion
const send = async (sendPrompt: SendPrompt) => {
```

File: bin/withdrawer.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    if (prompt.provider === "vault") {
+        keypair = await getKeypair(prompt.path!, prompt.key!);
+    } else {
+        keypair = Ed25519Keypair.fromSecretKey(decodeKeypair(prompt.keypair!));
+    }
+
+    return new RawSigner(keypair, provider);
+};
+
+const withdraw = async () => {
```
Comment (mikroskeem) [inline bin/withdrawer.ts:35]:
```suggestion
const withdraw = async (withdrawPrompt: Prompt) => {
```

File: bin/withdrawer.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import { getKeypair } from "../vault.ts";
+import { decodeKeypair } from "../utils.ts";
+import { provider, sendSuiObjects, withdrawStakeObjects } from "../sui.ts";
+
+type Prompt = {
+    provider: "vault" | "plain-text";
+    path: string | undefined;
+    key: string | undefined;
+    keypair: string | undefined;
+};
```
Comment (mikroskeem) [inline bin/withdrawer.ts:13]:
```suggestion
type Prompt = {
    provider: "vault" | "plain-text";
    path: string | undefined;
    key: string | undefined;
    keypair: string | undefined;
};

type VaultPrompt = Prompt & {
   provider: "vault";
   path: string;
   key: string;
};

type PlaintextPrompt = Prompt & {
    provider: "plain-text";
    keypair: string;
};

function isVaultPrompt(p: Prompt): p is VaultPrompt {
    return p.provider === "vault";
}

function isPlaintextPrompt(p: Prompt): p is PlaintextPrompt {
    return p.provider === "plain-text";
}
```

File: bin/withdrawer.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+let sendPrompt: SendPrompt;
+
+const getSigner = async (prompt: Prompt | SendPrompt): Promise<RawSigner> => {
+    let keypair: Ed25519Keypair;
+
+    if (prompt.provider === "vault") {
+        keypair = await getKeypair(prompt.path!, prompt.key!);
+    } else {
+        keypair = Ed25519Keypair.fromSecretKey(decodeKeypair(prompt.keypair!));
+    }
```
Comment (mikroskeem) [inline bin/withdrawer.ts:30]:
```suggestion
    if (isVaultPrompt(prompt)) {
        keypair = await getKeypair(prompt.path, prompt.key);
    } else if (isPlaintextPrompt(prompt)) {
        keypair = Ed25519Keypair.fromSecretKey(decodeKeypair(prompt.keypair));
    } else {
        throw new Error(`Unsupported provider "${prompt.provider}"`);
    }
```

File: bin/withdrawer.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+};
+
+const send = async () => {
+    const signer = await getSigner(sendPrompt);
+    const tx = await sendSuiObjects(signer, sendPrompt);
+    console.log(tx);
+};
+
+// deno-lint-ignore no-explicit-any
+const getPrompt = (): any => {
```
Comment (mikroskeem) [inline bin/withdrawer.ts:48]:
```suggestion
const getPrompt = <T>(): Promise<T> => {
```

File: bin/withdrawer.ts [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            after: async ({ confirmation }, next) => {
+                if (confirmation) {
+                    await next();
+                } else {
+                    // Reset the prompt...
+                    await next("provider");
+                }
+            },
+        },
+    ]);
```
Comment (mikroskeem) [inline bin/withdrawer.ts:101]:
```suggestion
    ]) as unknown as T;
```

File: bin/withdrawer.ts [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+};
+
+await new Command()
+    .name("sui-withdrawer")
+    .description("Easily withdraw Sui validator rewards")
+    .version("v0.0.1")
+    .command("withdraw", "Withdraw all staked Sui objects")
+    .action(async () => {
+        withdrawPrompt = await getPrompt();
+        await withdraw();
```
Comment (mikroskeem) [inline bin/withdrawer.ts:111]:
```suggestion
        await getPrompt<WithdrawPrompt>().then(withdraw);
```

File: bin/withdrawer.ts [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    })
+    .command("send", "Send Sui to a given address")
+    .arguments("<amount:number> <recipient:string>")
+    .action(async (_options, amount, recipient) => {
+        sendPrompt = await getPrompt();
+        sendPrompt.amount = amount;
+        sendPrompt.recipient = recipient;
+        await send();
```
Comment (mikroskeem) [inline bin/withdrawer.ts:119]:
```suggestion
        await getPrompt<SendPrompt>().then(send);
```

File: sui.ts [hunk 13]
```diff
@@ -1,57 +1,94 @@
...
 
-export const mergeSuiObjects = () => {
-    //
+    return transactions;
 };
 
-export const sendSuiObjects = () => {
-    //
+export const sendSuiObjects = async (signer: RawSigner, { amount, recipient }: SendPrompt): Promise<SuiTransactionBlockResponse> => {
+    let tx: SuiTransactionBlockResponse;
```
Comment (mikroskeem) [inline sui.ts:80]:
```suggestion
```

File: sui.ts [hunk 14]
```diff
@@ -1,57 +1,94 @@
...
+    const txb = new TransactionBlock();
+
+    const coin = txb.splitCoins(txb.gas, [txb.pure(amount * 1e9)]);
+    txb.transferObjects([coin], txb.pure(recipient));
+    try {
+        tx = await signer.signAndExecuteTransactionBlock({
```
Comment (mikroskeem) [inline sui.ts:86]:
```suggestion
        return await signer.signAndExecuteTransactionBlock({
```

File: sui.ts [hunk 15]
```diff
@@ -1,57 +1,94 @@
...
+            transactionBlock: txb,
+        });
+    } catch (e) {
+        throw new Error(`Failed to send ${amount} SUI to ${recipient}: ${e}`);
+    }
+
+    return tx;
```
Comment (mikroskeem) [inline sui.ts:93]:
```suggestion
```

File: vault.ts [hunk 16]
```diff
@@ -1,15 +1,41 @@
...
     namespace: config.VAULT_NAMESPACE,
     authentication,
 });
+
+export const getKeypair = async (path: string, key: string): Promise<Ed25519Keypair> => {
+    // We currently use environment variables for authentication.
+    try {
+        await client.login();
+    } catch (e) {
+        throw new Error(`Failed to login to Vault: ${e}`);
```
Comment (mikroskeem) [inline vault.ts:25]:
```suggestion
        throw new Error(`Failed to login to Vault`, {
            cause: e,
        });
```

File: vault.ts [hunk 17]
```diff
@@ -1,15 +1,41 @@
...
+    }
+
+    let keyData;
+    try {
+        keyData = await client.read(
+            createKVReadResponse(createKeySchema(key)),
+            path,
+        );
+    } catch (e) {
+        throw new Error(`Failed to read keypair from Vault: ${e}`);
```
Comment (mikroskeem) [inline vault.ts:35]:
```suggestion
        throw new Error(`Failed to read keypair from Vault`, {
            cause: e,
        });
```

File: sui.ts [hunk 18]
```diff
@@ -1,57 +1,94 @@
...
+    const txb = new TransactionBlock();
+
+    const coin = txb.splitCoins(txb.gas, [txb.pure(amount * 1e9)]);
+    txb.transferObjects([coin], txb.pure(recipient));
+    try {
+        tx = await signer.signAndExecuteTransactionBlock({
+            transactionBlock: txb,
+        });
+    } catch (e) {
+        throw new Error(`Failed to send ${amount} SUI to ${recipient}: ${e}`);
```
Comment (mikroskeem) [inline sui.ts:90]:
```suggestion
        throw new Error(`Failed to send ${amount} SUI to ${recipient}`, {
            cause: e,
        });
```

File: sui.ts [hunk 19]
```diff
@@ -1,57 +1,94 @@
...
         try {
-            await signer.signAndExecuteTransactionBlock({
+            tx = await signer.signAndExecuteTransactionBlock({
                 transactionBlock: txb,
             });
+            console.log(tx);
+            // We wait 2.5 seconds to allow the network process consecutive transactions.
+            await delay(3000);
         } catch (e) {
             throw new Error(`Failed to withdraw stake ${stake.stakedSuiId}: ${e}`);
```
Comment (mikroskeem) [inline sui.ts:71]:
```suggestion
            throw new Error(`Failed to withdraw stake ${stake.stakedSuiId}`, {
                cause: e,
            });
```

**Review by mikroskeem — CHANGES_REQUESTED** [review]

**Review by mikroskeem — APPROVED** [review]

**Pull request merged**